jplag.frontend-utils/src/main/java/de/jplag/Token.java

```diff
@@ -5,46 +5,45 @@
  * What types of tokens there are depends on the specific language, meaning JPlag does not enforce a specific token set.
  * The language parsers decide what is a token and what is not.
  */
-public abstract class Token {
+public class Token {
     /** Indicates that the requested field has no value. */
     public static final int NO_VALUE = -1;
 
     private int line;
     private int column;
     private int length;
     private String file;
-
-    protected int type;
+    private TokenType type;
 
     /**
-     * Creates a token without information about the column or the length of the token in the line.
-     * @param type is the token type.
+     * Creates a token of type {@link SharedTokenType#FILE_END FILE_END} without information about line, column, and length.
      * @param file is the name of the source code file.
-     * @param line is the line index in the source code where the token resides. Cannot be smaller than 1. For
-     * {@link TokenConstants#FILE_END FILE_END} it is automatically set to {@link #NO_VALUE}.
      */
-    public Token(int type, String file, int line) {
-        this.type = type;
-        this.file = file;
-        if (type == TokenConstants.FILE_END) {
-            this.line = NO_VALUE;
-        } else {
-            this.line = line > 0 ? line : 1;
-        }
+    public static Token fileEnd(String file) {
+        return new Token(SharedTokenType.FILE_END, file, NO_VALUE, NO_VALUE, NO_VALUE);
     }
 
     /**
      * Creates a token with column and length information.
      * @param type is the token type.
      * @param file is the name of the source code file.
-     * @param line is the line index in the source code where the token resides. Cannot be smaller than 1.
+     * @param line line is the line index in the source code where the token resides. Cannot be smaller than 1. For
+     * {@link SharedTokenType#FILE_END FILE_END} it is automatically set to {@link #NO_VALUE}.
      * @param column is the column index, meaning where the token starts in the line.
      * @param length is the length of the token in the source code.
      */
-    public Token(int type, String file, int line, int column, int length) {
-        this(type, file, line);
-        this.column = column;
-        this.length = length;
+    public Token(TokenType type, String file, int line, int column, int length) {
+        this.type = type;
+        this.file = file;
+        if (type == SharedTokenType.FILE_END) {
+            this.line = NO_VALUE;
+            this.column = NO_VALUE;
+            this.length = NO_VALUE;
```

Comment:
Implemented in https://github.com/jplag/JPlag/pull/623/commits/2109ab4be9d69907078c1f5cbf2cc9cd113ee6a7.
I decided against using assertions in production code, as this only affects the rendering of the matches but not the similarity score computation, and instead extended the documentation and logger warn messages.

Reply:
I am not sure if a warning is a good choice, as for JPlag, they are mainly warnings for the user and not the developer. If we say the token indexing is one-based, why not just throw an `IllegalArgumentException` whenever the value of line or columns is < 1 and not -1 (e.g. `line < 1 && line != NO_VALUE`). This shows the language developer that he is clearly misusing the token class, and the tokens will break the viewer. I think a warning is a bit too soft here.

Reply:
I can see arguments for both sides, however, I am still in favor of the Exception. Maybe @dfuchss can be the tie-breaker here so we can move forward and merge this PR.

Reply:
I important question is whether this leads to visualization bugs or calculation bugs.
If the only problem is a visualization bug, I would be ok with an error/warning that says "Illegal Tokens found while generation. Please contact the JPlag Team. This may lead to visualization bugs."

If the problem might lead to errors in the calculation in the Core, we have to throw an Error.

Do you have some insights here? **If not, I would argue for an exception because it would be unclear what could happen else.**

Reply:
It for sure only affects the visualization. For the calculation, only the token type is relevant.